[9.0][IMP][base_custom_info] New types and access rules system.

yajo opened the pull request:
Now you can define properties types, and access rules are inherited from the model/record linked to the custom info record.

@Tecnativa

yajo committed:
[CHG] base_custom_info: Relicense to LGPL with all authors' consent

yajo committed:
[IMP]base_custom_info: New types and access rules system.

* Now you can define properties types, and access rules are inherited from the model/record
  linked to the custom info record.
* Simplified version of computed value.
* Implement for res.partner.
* Add tests and fix bugs discovered in the meantime.
* Allow to disable partner custom info tab, and custom info menu.
* All of it can be set within general settings.
* Now, by default, this module does not display custom info for partners unless in demo mode.
  Better fit for a base module.
* You can disable the top menu entry too if it disturbs you, or enable it for everybody.
* Give a special form when editing in partner custom info tab.
* Sortable properties.
* Sort values at onchange time.
* Improve performance in onchange.
* Split in several model files.

pedrobaeza commented:
Texts fields should be translatable. Can that be configurable in the definition? I don't think so, but maybe we can make some hack in the method `fields_get` or `fields_view_get`.

If not, we should define then the field as always translatable but include anyway a flag that allows to specify this translatable possibility. If you don't mark this flag, then at write time, you have to synchronize by code all the translations to the current one, so it behaves as if the field is not translatable in the background.

pedrobaeza commented:
Why the test coverage drop?

pedrobaeza commented on base_custom_info/README.rst:
```diff
@@ -26,7 +26,7 @@ expected for a given record.
 To define a template, you need to:
 
-* Go to *Settings > Custom Info > Templates*.
+* Go to *Custom Info > Templates*.
```
Is this a top level menu?

pedrobaeza replied:
Isn't possible to put inside another top level menu? (we have enough top level menus already)

yajo replied:
I realized that if you have access to setting the templates and values, you should be able to manage them from their menu more comfortably, no need to be an admin for that. Given in v9 the trend is to have more dedicated menus, I'm just following it. Besides, this base module isn't related to any other by default.

I guess the increase of top level menus comes from the enterprise version, where they are beautiful icons in a smartphone-like home panel. After all, the problem is that the backend community interface is way old-fashioned. Nobody will care about them if some day OCA creates a good alternative :yum:

pedrobaeza replied:
OK, but after this day comes, we have a lot of them... Anyway, if you don't find any parent menu, let's keep it this way. But we should provide then record rules to avoid one person without product access that modifies custom info profiles.

pedrobaeza replied:
Forget about the record rule. You have answered me in other comment.

lasley replied:
Note- mobile ready interface is scheduled in one of our sprints, so coming soon 😄

pedrobaeza commented on base_custom_info/models/custom_info.py:
```diff
@@ -85,7 +119,126 @@ class CustomInfoValue(models.Model):
+        """Get the value as a string, from the original field."""
+        for s in self:
+            try:
+                s.value = s[s.field_name]
```
Use `getattr(s, field_name, False)` and avoid the try sentence

lasley replied:
Wouldn't this be `s.get(` for a dict lookup instead of `getattr`?

lasley replied:
However @Yajo is actually being more Pythonic + efficient here IMO due to utilization of the else in the try block. In order to implement that, a new conditional would need to be created to eval vs. just going one way or the other based on a bubbled exception.

pedrobaeza commented on base_custom_info/models/custom_info.py:
```diff
@@ -66,6 +76,30 @@ class CustomInfoProperty(models.Model):
         string="Property Values")
+    field_type = fields.Selection(
+        selection=[
+            ("str", "Text"),
```
Why not using the same wording as the field names (Char, Float, Boolean...)?

yajo replied:
Simplifies line 261

pedrobaeza commented on base_custom_info/security/ir.model.access.csv:
```diff
@@ -1,7 +1,5 @@
-"access_custom_info_property_sale_manager","custom.info.template.line.salemanager","model_custom_info_property","base.group_system",1,1,1,1
-"access_custom_info_value_sale_manager","custom.info.value.salemanager","model_custom_info_value","base.group_system",1,1,1,1
+id,name,model_id:id,group_id:id,perm_read,perm_write,perm_create,perm_unlink
+access_template,Access custom info templates,model_custom_info_template,,1,1,1,1
```
Why allowing everyone to modify this?

yajo replied:
I realized that it makes no sense to have access rules for this model. Custom information is always linked to another model, that is the one with the access rules, so you will notice that I oveloaded each model's `check_access_rule` method to make them inherit the access rules of their linked model.

This way, if you can edit a product.template that has a custom.info.value, you can edit their values. If you cannot, you can't also edit their values. All is automatic and sensible.

pedrobaeza replied:
OK, understood.

yajo commented:
About coverage drop, I think it's because this module had no tests and it just got bigger.

I'd need some model to implement the `custom.info` mixin to test this. Do you have any idea on how to make the tests?

Can I declare a new model inside a test? Or do I create a transient model just for testing purposes, that does not appear anywhere else? But it would exist even in production databases...

pedrobaeza commented:
Make it over res.users model.

pedrobaeza commented:
Or even over res.partner. They are both models that are in the base module.

yajo commented:
Good idea! :bulb:

pedrobaeza commented:
Change spanish translation of the menu "Custom info" to "Inf. personalizada" to shorten the possible the menu width

pedrobaeza commented on base_custom_info/views/custom_info_value_view.xml:
```diff
@@ -1,19 +1,53 @@
     <field name="name">base.custom.info.value.tree</field>
     <field name="model">custom.info.value</field>
     <field name="arch" type="xml">
         <tree string="Custom Property Values">
```
Put `create="0" delete="0"` to avoid modifications on the populated list

yajo replied:
Actually I thought that letting users put additional properties could help under some circumnstances. What do you think?

pedrobaeza replied:
I don't think so. This will lead to conflicts (lines that doesn't belong to the same model), and it also break the concept of information template. If you need something extra, make another profile. The only problem I see with this that the same property can't be reused for several templates. Or it can be?

yajo replied:
Not right now. I can add that to known issues.

yajo replied:
BTW, this inability simplifies one algorithm. I'll commit soon.

pedrobaeza commented on base_custom_info/views/custom_info_value_view.xml:
```diff
@@ -1,19 +1,53 @@
+                        name="value_bool"
+                        attrs="{'invisible': [('field_type', '!=', 'bool')]}"/>
+                    <field
+                        name="value_id"
```
put `widget="selection"`

yajo replied:
Done.

yajo commented:
This module should be ready, but there's a bug that me and @pedrobaeza found, that seems to affect v9 onchange methods.

When you choose a template from the partner form, it fills the values, but properties remain empty:

![captura de pantalla de 2016-08-04 17-47-37](https://cloud.githubusercontent.com/assets/973709/17408464/936bf490-5a6b-11e6-857e-b0024a6990de.png)

yajo commented:
I honestly don't know why, but it seems to work now, after adding custom tree and form views inside the field.

Does anybody know how to disable the button "save & new" in the modal form?

lasley commented on base_custom_info/models/custom_info.py:
```diff
@@ -2,9 +2,10 @@
+from openerp.tools.safe_eval import safe_eval
 
 
 class CustomInfoModelLink(models.AbstractModel):
```
~~Can we get these models in separate files please? [Relevant guideline](https://github.com/OCA/maintainer-tools/blame/master/CONTRIBUTING.md#L941)~~ Feel free to ignore me here, my bad didn't realize this wasn't a MIG for a second

lasley commented on base_custom_info/models/custom_info.py:
```diff
@@ -85,7 +119,126 @@ class CustomInfoValue(models.Model):
+            elif s.field_type == "id":
+                parts = (p.strip() for p in value.split(","))
+                for part in parts:
+                    value = self.env["custom.info.option"].search(
```
Looks like this will be a recordset, but all other `value` vars are simply a value. Seems like it works functionally though, just wanted to make sure it was on purpose.

yajo replied:
Yes, see that this time you are in the "selection" type of field, which has a m2o from custom.info.option.

lasley replied:
Cool thanks for the clarification

lasley commented:
@Yajo does `action_buttons="false"` in the form tag declaration for the modal work? Admittedly never used it, but I found [this code](https://github.com/odoo/odoo/blob/7682760dcad232829fbb5d4221f66a1ebb9d6c91/addons/web_view_editor/static/src/js/view_editor.js#L57) the other day, which seems to elude to what we want.

yajo commented:
@lasley Seems impossible. [I'd have to add the `disable_multiple_selection` somewhere](https://github.com/odoo/odoo/blob/9.0/addons/web/static/src/js/views/form_common.js#L854), but [options the form gets are hardcoded](https://github.com/odoo/odoo/blob/9.0/addons/web/static/src/js/views/form_relational_widgets.js#L1102-L1120).

yajo commented:
This should be ready to merge now.

lasley commented:
Lame about the hardcoded form attrs. LGTM thanks @Yajo 👍

pedrobaeza commented on base_custom_info/models/custom_info.py:
```diff
@@ -119,7 +119,7 @@ class CustomInfoValue(models.Model):
     _name = "custom.info.value"
     _rec_name = 'value'
-    _order = "property_id"
+    _order = "name"
```
I think this should be ordered by a sequence defined in the template (not sortable at user level).

yajo replied:
I like that.

yajo commented:
I rebased to fix conflicts, and took the time to squash some commits. Sorry reviewers :sweat_smile:. Notice that the really new 2 commits are the last 2.

gurneyalex commented:
@Yajo  the 9.0 travis build was recently fixed, can you rebase your PR?

gurneyalex commented on base_custom_info/README.rst:
```diff
@@ -1,6 +1,6 @@
+    :alt: License: LGPL-3
 
 ================
 Base Custom Info
```
Could this section be improved. I miss a description of why / when I could need this module. 

It fhis is a "technical dependency", referring the reader to other modules depending on this one would be a nice touch.

yajo replied:
Fixed in https://github.com/OCA/server-tools/pull/492/commits/be58eb8e63b43e2c7b176662b920c4281002da26.

gurneyalex commented on base_custom_info/README.rst:
```diff
@@ -26,7 +37,7 @@ expected for a given record.
 
 To define a template, you need to:
 
```
I would like a definition of "template" in this context.

yajo replied:
Fixed in https://github.com/OCA/server-tools/pull/492/commits/be58eb8e63b43e2c7b176662b920c4281002da26.

gurneyalex commented on base_custom_info/README.rst:
```diff
@@ -26,7 +37,7 @@ expected for a given record.
-* Go to *Settings > Custom Info > Templates*.
+* Go to *Custom Info > Templates*.
 * Create one.
 * Add some *Properties* to it.
```
I would like a definition of "property" in this context. This is important as "property" has a special meaning in Odoo, and I'm not sure these are the same (i.e. fields with a company dependent value)

yajo replied:
Fixed in https://github.com/OCA/server-tools/pull/492/commits/be58eb8e63b43e2c7b176662b920c4281002da26.

pedrobaeza commented on base_custom_info/README.rst:
```diff
@@ -6,8 +6,117 @@
+Does this work with any model?
+------------------------------
+
+Yes and not.
```
s/not/no

pedrobaeza commented on base_custom_info/README.rst:
```diff
@@ -6,8 +6,117 @@
+Yes, because this is a base module that provides the tools to make this work
+with any model.
+
+Not, because, although the tools are provided, they are only applied to the
```
s/Not/no

pedrobaeza commented on base_custom_info/README.rst:
```diff
@@ -6,8 +6,117 @@
+
+So, if you want to apply this to other models, you will have to develop a
+little additional addon that depends on this one. If you are a developer, see
+an example in the ``product_custom_info`` addon.
```
s/see an example/see section _Developer_ and an example

sergio-teruel commented on base_custom_info/models/custom_info_value.py:
```diff
@@ -0,0 +1,224 @@
+    )
+    category_id = fields.Char(
+        related="property_id.category_id",
+        store=True,
```
related="property_id.category_id.name" ? if category_id as a Char is the name.

yajo replied:
Actually it's the field type what's wrong, but thanks for the point.

sergio-teruel commented on base_custom_info/models/custom_info.py:
```diff
@@ -48,3 +48,13 @@ def unlink(self):
+        return self.env["custom.info.value"].search([
+            ("model", "=", self._name),
+            ("res_id", "in", self.ids),
+            ("property_ids", "in", properties.ids),
```
("property_id", "in", ....)

yajo commented:
Now all should go :green_heart: again :blush:

lasley commented:
It was red on `odoo/odoo` but I reran and it's 🍏  now. Code looks good too 👍

gaelTorrecillas commented:
hi @Yajo 

https://github.com/OCA/server-tools/pull/492/files#diff-79eae0ef2d232914029f683730cb993eR23

why this field is read only? we can't select the model so we can't validate the new template.

yajo commented:
This is because the templates are expected to be created from the owner model, when selecting it > create new, so you cannot change them later.

guewen commented on base_custom_info/models/custom_info_category.py:
```diff
@@ -0,0 +1,34 @@
+            except Exception as last:
+                pass
+        if last:
+            raise last
```
Why not raise directly?

``` python
for prop in self.mapped("property_ids"):
    prop.check_access_rule(operation)
    break
```

yajo replied:
The docstring says that you need to access at least one property. Imagine you can access the 2nd one in the recordset. Doing what you say would forbid access when the 1st one fails. However there's one failure, we should `return` instead of `break`ing.

guewen replied:
Got it, then yeah, better with return :)

guewen commented on base_custom_info/models/custom_info_option.py:
```diff
@@ -0,0 +1,38 @@
+            except Exception as last:
+                pass
+        if last:
+            raise last
```
Why not raise directly?

``` python
for prop in self.mapped("property_ids"):
    prop.check_access_rule(operation)
    break
```

guewen commented on base_custom_info/models/custom_info_template.py:
```diff
@@ -0,0 +1,66 @@
+        for s in self:
+            if s.model not in self.env:
+                raise ValidationError(_("Model does not exist."))
+            with self.env.norecompute():
```
Could you add a comment explaining why you did need to use this context manager?

yajo replied:
I think the function's docstring explains it.

guewen commented:
Thanks :) :+1:

guewen commented on base_custom_info/models/custom_info_option.py:
```diff
@@ -0,0 +1,38 @@
+        for prop in self.mapped("property_ids"):
+            try:
+                prop.check_access_rule(operation)
+                break
```
So you should have a `return` here too

guewen commented:
Seems there is clash with the `database_cleanup` addon :-(
Is it something you think we can resolve or perhaps should we separate the addons in different test runs?

``` python
2016-09-01 08:10:53,921 24202 ERROR openerp_test openerp.addons.database_cleanup.tests.test_database_cleanup: ERROR

2016-09-01 08:10:53,924 24202 INFO openerp_test openerp.addons.database_cleanup.tests.test_database_cleanup: ======================================================================

2016-09-01 08:10:53,924 24202 ERROR openerp_test openerp.addons.database_cleanup.tests.test_database_cleanup: ERROR: test_database_cleanup (openerp.addons.database_cleanup.tests.test_database_cleanup.TestDatabaseCleanup)

2016-09-01 08:10:53,924 24202 ERROR openerp_test openerp.addons.database_cleanup.tests.test_database_cleanup: Traceback (most recent call last):

2016-09-01 08:10:53,924 24202 ERROR openerp_test openerp.addons.database_cleanup.tests.test_database_cleanup: `   File "/home/travis/build/OCA/server-tools/database_cleanup/tests/test_database_cleanup.py", line 77, in test_database_cleanup

2016-09-01 08:10:53,924 24202 ERROR openerp_test openerp.addons.database_cleanup.tests.test_database_cleanup: `     purge_tables.purge_all()

2016-09-01 08:10:53,925 24202 ERROR openerp_test openerp.addons.database_cleanup.tests.test_database_cleanup: `   File "/home/travis/odoo-9.0/openerp/api.py", line 248, in wrapper

2016-09-01 08:10:53,925 24202 ERROR openerp_test openerp.addons.database_cleanup.tests.test_database_cleanup: `     return new_api(self, *args, **kwargs)

2016-09-01 08:10:53,925 24202 ERROR openerp_test openerp.addons.database_cleanup.tests.test_database_cleanup: `   File "/home/travis/build/OCA/server-tools/database_cleanup/models/purge_wizard.py", line 51, in purge_all

2016-09-01 08:10:53,925 24202 ERROR openerp_test openerp.addons.database_cleanup.tests.test_database_cleanup: `     self.mapped('purge_line_ids').purge()

2016-09-01 08:10:53,925 24202 ERROR openerp_test openerp.addons.database_cleanup.tests.test_database_cleanup: `   File "/home/travis/odoo-9.0/openerp/api.py", line 248, in wrapper

2016-09-01 08:10:53,925 24202 ERROR openerp_test openerp.addons.database_cleanup.tests.test_database_cleanup: `     return new_api(self, *args, **kwargs)

2016-09-01 08:10:53,925 24202 ERROR openerp_test openerp.addons.database_cleanup.tests.test_database_cleanup: `   File "/home/travis/build/OCA/server-tools/database_cleanup/models/purge_tables.py", line 63, in purge

2016-09-01 08:10:53,925 24202 ERROR openerp_test openerp.addons.database_cleanup.tests.test_database_cleanup: `     "DROP TABLE %s", (IdentifierAdapter(line.name),))

2016-09-01 08:10:53,926 24202 ERROR openerp_test openerp.addons.database_cleanup.tests.test_database_cleanup: `   File "/home/travis/odoo-9.0/openerp/sql_db.py", line 141, in wrapper

2016-09-01 08:10:53,926 24202 ERROR openerp_test openerp.addons.database_cleanup.tests.test_database_cleanup: `     return f(self, *args, **kwargs)

2016-09-01 08:10:53,926 24202 ERROR openerp_test openerp.addons.database_cleanup.tests.test_database_cleanup: `   File "/home/travis/odoo-9.0/openerp/sql_db.py", line 220, in execute

2016-09-01 08:10:53,926 24202 ERROR openerp_test openerp.addons.database_cleanup.tests.test_database_cleanup: `     res = self._obj.execute(query, params)

2016-09-01 08:10:53,926 24202 ERROR openerp_test openerp.addons.database_cleanup.tests.test_database_cleanup: ` InternalError: cannot drop table custom_info_template because other objects depend on it

2016-09-01 08:10:53,926 24202 ERROR openerp_test openerp.addons.database_cleanup.tests.test_database_cleanup: ` DETAIL:  constraint res_partner_custom_info_template_id_fkey on table res_partner depends on table custom_info_template
```

yajo commented:
It seems like that addon wants to delete the `custom_info_template` table, which I honestly have no idea why... All I can say is I'm happy it can't. :laughing:

Any idea?

yajo commented:
Cool, now you can have required fields and default values, and view definitions for submodules are much thinner.

I still have to implement recursive templates, please mark as WIP.

yajo committed:
[IMP] base_custom_info: Recursive templates.

- Select an option with an additional template and get it filled on the owner.
- Drop support for onchange, too many bugs to workaround.
- Improve demo data.
- Improve tests.
- Almost cool material icon, this is an app now!
- Fun pics.

yajo commented:
As this keeps getting more awesome, more weird odoo bugs keep hitting it. I'm applying workarounds for https://github.com/odoo/odoo/pull/13439, https://github.com/odoo/odoo/pull/11042, https://github.com/odoo/odoo/issues/13076, https://github.com/odoo/odoo/pull/13493, and, worst of all, https://github.com/odoo/odoo/pull/13480 and https://github.com/odoo/odoo/pull/13082.

yajo commented:
What's up with travis? :confused:

rafaelbn commented:
Well, travis is red but we have been working for many months with this PR... @Yajo ?

yajo commented:
That's because travis failures are from other addons, I rebased now, let's see if some get fixed.

lasley commented on base_custom_info/models/custom_info.py:
```diff
@@ -124,3 +58,64 @@ def unlink(self):
+            to_add = props_good - props_enabled
+            to_rm = props_enabled - props_good
+            # Remove remaining properties
+            # HACK https://github.com/odoo/odoo/pull/13480
```
This is merged, good to remove?

lasley commented on base_custom_info/models/custom_info.py:
```diff
@@ -1,121 +1,55 @@
+        compute="_compute_dirty_templates",
+    )
+
+    # HACK https://github.com/odoo/odoo/pull/11042
```
Closed as no-fix. Probably good to remove the comment?

lasley commented on base_custom_info/models/custom_info.py:
```diff
@@ -1,121 +1,55 @@
+                "dirty_templates").action_custom_info_templates_fill()
+        return res
 
+    # HACK https://github.com/odoo/odoo/pull/11042
```
Closed as no-fix. Probably good to remove the comment?

lasley commented on base_custom_info/models/custom_info.py:
```diff
@@ -124,3 +58,64 @@ def unlink(self):
             info_values.unlink()
         return res
+
+    @api.one
```
Deprecated, please switch to multi

lasley commented on base_custom_info/models/custom_info.py:
```diff
@@ -124,3 +58,64 @@ def unlink(self):
+    @api.multi
+    def all_custom_info_templates(self):
+        """Get all custom info templates involved in these owners."""
+        return (self.mapped("custom_info_template_id") |
```
Bitwise or, on a recordset?

yajo replied:
This works as with normal python sets, joining all results and removing duplicates.

lasley replied:
TIL

lasley commented on base_custom_info/models/custom_info_property.py:
```diff
@@ -0,0 +1,113 @@
+        self.mapped("template_id").check_access_rule(operation)
+        return super(CustomInfoProperty, self).check_access_rule(operation)
+
+    @api.one
```
Deprecated

lasley commented on base_custom_info/models/custom_info_property.py:
```diff
@@ -0,0 +1,113 @@
+                    self.default_value, self.field_type, self)
+            except ValueError:
+                selection = dict(
+                    self._fields["field_type"].get_description(self.env)
```
Missing a comma between this and next arg. Although why `dict()` instead of `{}`?

lasley commented on base_custom_info/models/custom_info_value.py:
```diff
@@ -0,0 +1,302 @@
+        if "value" in vals:
+            self.env.context.skip_required = True
+        result = super(CustomInfoValue, self).create(vals)
+        # HACK https://github.com/odoo/odoo/pull/11042
```
closed nofix

lasley commented on base_custom_info/models/custom_info_value.py:
```diff
@@ -0,0 +1,302 @@
+                .action_custom_info_templates_fill()
+        return result
+
+    # HACK https://github.com/odoo/odoo/pull/11042
```
closed nofix

lasley commented on base_custom_info/models/custom_info_value.py:
```diff
@@ -0,0 +1,302 @@
+            s[s.field_name] = self._transform_value(
+                s.value, s.field_type, s.property_id)
+
+    @api.one
```
deprecated

lasley commented on base_custom_info/models/custom_info_value.py:
```diff
@@ -0,0 +1,302 @@
+    @api.onchange("property_id")
+    def _onchange_property_set_default_value(self):
+        """Load default value for this property."""
+        for record in self:
```
No need to rec loop in an onchange

lasley reviewed (changes requested):
Thanks @Yajo - comments inline

PS big 👍  to the images

pedrobaeza committed:
[IMP] base_custom_info: Make backend UI work

yajo commented on base_custom_info/models/custom_info_value.py:
```diff
@@ -266,10 +206,10 @@ def _transform_value(self, value, format_, properties=None):
-            value.ensure_one()
+            value = self.env["custom.info.option"].search([
+                ("property_ids", "in", properties.ids),
+                ("name", "ilike", u"%{}%".format(value)),
```
These percents do nothing, right? 🤔 (It's not `=ilike`)

pedrobaeza replied:
Uhm, can be, I wasn't sure and put that for assuring.

yajo replied:
[Indeed.](https://www.odoo.com/documentation/9.0/reference/orm.html#domains)

Yajo commented:
I cannot install now in runbot:

```
Odoo Server Error

Traceback (most recent call last):
  File "/.repo_requirements/odoo/openerp/http.py", line 648, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/.repo_requirements/odoo/openerp/http.py", line 685, in dispatch
    result = self._call_function(**self.params)
  File "/.repo_requirements/odoo/openerp/http.py", line 321, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/.repo_requirements/odoo/openerp/service/model.py", line 118, in wrapper
    return f(dbname, *args, **kwargs)
  File "/.repo_requirements/odoo/openerp/http.py", line 314, in checked_call
    result = self.endpoint(*a, **kw)
  File "/.repo_requirements/odoo/openerp/http.py", line 964, in __call__
    return self.method(*args, **kw)
  File "/.repo_requirements/odoo/openerp/http.py", line 514, in response_wrap
    response = f(*args, **kw)
  File "/root/odoo-9.0/addons/web/controllers/main.py", line 892, in call_button
    action = self._call_kw(model, method, args, {})
  File "/root/odoo-9.0/addons/web/controllers/main.py", line 880, in _call_kw
    return getattr(request.registry.get(model), method)(request.cr, request.uid, *args, **kwargs)
  File "/.repo_requirements/odoo/openerp/api.py", line 250, in wrapper
    return old_api(self, *args, **kwargs)
  File "/.repo_requirements/odoo/openerp/addons/base/module/module.py", line 459, in button_immediate_install
    return self._button_immediate_function(cr, uid, ids, self.button_install, context=context)
  File "/.repo_requirements/odoo/openerp/api.py", line 250, in wrapper
    return old_api(self, *args, **kwargs)
  File "/.repo_requirements/odoo/openerp/addons/base/module/module.py", line 534, in _button_immediate_function
    registry = openerp.modules.registry.RegistryManager.new(cr.dbname, update_module=True)
  File "/.repo_requirements/odoo/openerp/modules/registry.py", line 386, in new
    openerp.modules.load_modules(registry._db, force_demo, status, update_module)
  File "/.repo_requirements/odoo/openerp/modules/loading.py", line 340, in load_modules
    loaded_modules, update_module)
  File "/.repo_requirements/odoo/openerp/modules/loading.py", line 239, in load_marked_modules
    loaded, processed = load_module_graph(cr, graph, progressdict, report=report, skip_modules=loaded_modules, perform_checks=perform_checks)
  File "/.repo_requirements/odoo/openerp/modules/loading.py", line 159, in load_module_graph
    _load_data(cr, module_name, idref, mode, kind='demo')
  File "/.repo_requirements/odoo/openerp/modules/loading.py", line 98, in _load_data
    tools.convert_file(cr, module_name, filename, idref, mode, noupdate, kind, report)
  File "/.repo_requirements/odoo/openerp/tools/convert.py", line 843, in convert_file
    convert_csv_import(cr, module, pathname, fp.read(), idref, mode, noupdate)
  File "/.repo_requirements/odoo/openerp/tools/convert.py", line 907, in convert_csv_import
    raise Exception(_('Module loading %s failed: file %s could not be processed:\n %s') % (module, fname, warning_msg))
Exception: Module loading base_custom_info failed: file base_custom_info/demo/custom.info.template.csv could not be processed:
 Line 1 : null value in column "model_id" violates not-null constraint
DETAIL:  Failing row contains (34, null, 2017-02-13 08:21:16.225503, Smart partners, 1, 1, 2017-02-13 08:21:16.225503).
```

sergio-teruel committed:
[FIX] base_custom_info: fix demo template csv

rafaelbn commented:
@lasley could you review here last changes, your revision is blocking, thanks!
@Yajo please check travis, is failing

yajo commented:
I just tested this and it works like a charm now! I hope I can fix Travis soon & merge, awesome pals.

dreispt commented:
I tried a Travis rebuild 🤞

yajo committed:
[FIX][base_custom_info] Make tests work

yajo commented:
It should go 💚 now.
